### PR TITLE
Revert "Revert "[core][experimental] Avoid serialization for data passed between two tasks on the same actor""

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

The fix is at https://github.com/ray-project/ray/pull/45937/commits/75ac14f1ab70ec7d6ea24c6c85aa8378c1d1356a. After #45786, we no longer use None to signify the driver process. Instead, we create a DAGDriverProxyActor.

I just cherry-picked the merged commit as this PR's first commit and added the fix in the second commit. Therefore, we only need to focus on the second commit.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_channel.py [hunk 1]
```diff
@@ -591,6 +591,195 @@ def close(self):
...
+
+    @ray.remote(num_cpus=1)
+    class Actor:
+        def __init__(self):
+            pass
+
+        def pass_channel(self, channel):
+            self._chan = channel
+
+        def create_multi_channel(self, writer, readers):
```
Comment (ruisearch42) [inline python/ray/tests/test_channel.py:737]: nit: create_composite_channel

**Review by stephanie-wang — APPROVED** [review]

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 2]
```diff
@@ -542,11 +542,19 @@ def write(self, value: Any):
     def begin_read(self) -> Any:
         self.ensure_registered_as_reader()
         actor_id = ray.get_runtime_context().get_actor_id()
+        if actor_id is None:
+            # The reader is the driver process.
+            # Use the actor ID of the DAGDriverProxyActor.
+            actor_id = self._get_actor_id(self._readers[0])
```
Comment (ruisearch42) [inline python/ray/experimental/channel/shared_memory_channel.py:548]: wrap this into a method?

Comment (stephanie-wang) [reply python/ray/experimental/channel/shared_memory_channel.py:548]: Actually, can you also assert here that len(self._readers) == 1? Want to make sure that we are picking the right reader.

File: python/ray/tests/test_channel.py (hunk 1 above)
Comment (kevin85421) [reply python/ray/tests/test_channel.py]: Updated https://github.com/ray-project/ray/pull/45937/commits/32d14f80166de400ca459b4c7d1b8b7c49dd9771

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 2 above)
Comment (kevin85421) [reply python/ray/experimental/channel/shared_memory_channel.py]: Updated https://github.com/ray-project/ray/pull/45937/commits/32d14f80166de400ca459b4c7d1b8b7c49dd9771

**Review by ruisearch42 — APPROVED** [review]
LGTM. Thanks for the change.

File: python/ray/dag/compiled_dag_node.py [hunk 3]
```diff
@@ -45,7 +45,7 @@ def do_allocate_channel(
 
     Args:
         readers: The actor handles of the readers.
-        buffer_size_bytes: The maximum size of messages in the channel.
```
Comment (jjyao) [inline python/ray/dag/compiled_dag_node.py:48]: typo? type?

Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:48]: No. `typ` follows the naming convention of Ray codebase.

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 4]
```diff
@@ -507,11 +507,21 @@ def __init__(
                 actor_id = self._get_actor_id(reader)
                 self._channel_dict[actor_id] = remote_channel
 
-    def _get_actor_id(self, reader: Optional[ray.actor.ActorHandle]) -> str:
-        if reader is None:
-            return None
+    def _get_actor_id(self, reader: ray.actor.ActorHandle) -> str:
         return reader._actor_id.hex()
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/experimental/channel/shared_memory_channel.py:511]: I think it is OK to make it a function. I have some experience where I compare `reader._actor_id` with other strings. I think it is common for people to forget to call `hex()`.

File: python/ray/dag/compiled_dag_node.py (hunk 3 above)
Comment (jjyao) [reply python/ray/dag/compiled_dag_node.py:48]: ok, seems we only call it `typ` in ADAG codebase tho.

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 5]
```diff
@@ -440,3 +447,118 @@ def close(self) -> None:
...
+    def _get_self_actor_id(self) -> str:
+        """
+        Get the actor ID of the current process. If the current process is the driver,
+        use the actor ID of the DAGDriverProxyActor.
+        """
+        actor_id = ray.get_runtime_context().get_actor_id()
+        if actor_id is None:
+            # The reader is the driver process.
+            # Use the actor ID of the DAGDriverProxyActor.
+            actor_id = self._get_actor_id(self._readers[0])
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:522]:
```suggestion
            assert len(self._readers) == 1
            actor_id = self._get_actor_id(self._readers[0])
```

**Pull request merged**